src/content/blog/astro-4100.mdx

```diff
@@ -0,0 +1,208 @@
+
+To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:
+
+```sh
+# Recommended:
+npx @astrojs/upgrade
+
+# Manual:
+npm install astro@latest
+pnpm upgrade astro --latest
+yarn upgrade astro --latest
+```
+
+## Experimental: astro:env
+
+Astro 4.10 introduces a new experimental built-in module, `astro:env`, to allow easier use of environment variables.
+
+Environment variables allow you to configure your app with different values in different environments. But this comes with a great deal of complexity:
+
+- Some variables are needed in the client and some only on the server.
+- Server variables are often __secrets__, things like API keys that you absolutely do not want exposed in the client, and you also do not want inlined into the server build either (which can be viewed by anyone with access to the build output).
+- Some variables are __required__ for your app to function at all; whereas others are optional enhancements.
+- Variables can be defined in your shell, in a `.env` file, or in [build config](https://vitejs.dev/config/#using-environment-variables-in-config).
+- Runtimes such as [Cloudflare](https://docs.astro.build/en/guides/integrations-guide/cloudflare/) and [Deno](https://deno.com/) have different APIs for reading variables, creating a dev/prod difference you need to deal with.
+
+We built `astro:env` to provide more control and structure over environment variables. Here's how you might define a schema for them in your config:
+
+```js
+import { defineConfig, envField } from 'astro/config';
+
+export default defineConfig({
+  experimental: {
+    env: {
+      schema: {
+        API_PORT: envField.number({
+          context: 'server',
+          access: 'secret',
+          default: 7000
+        }),
+        PUBLIC_DASHBOARD_V2: envField.boolean({
+          context: 'server',
+          access: 'public',
+          default: false
+        }),
+      }
+    }
+  }
+})
+```
+
+Once defined you can use your variables by importing them from the `astro:env/server` and `astro:env/client` modules like so:
+
+```js
+import { PUBLIC_DASHBOARD_V2, getSecret } from 'astro:env/server';
+
+if (PUBLIC_DASHBOARD_V2) {
+	const API_PORT = getSecret("API_PORT") // number
+	await fetch(`https://my-secret-api.com:${API_PORT}/v2`)
+}
+```
+
+Client-side `astro:env/client` can be used in components, scripts, or anywhere else you run client code. For example:
+
+```jsx
+import { SOME_FEATURE_FLAG } from 'astro:env/client';
+
+export default function() {
+  return (
+    <section>
+      { SOME_FEATURE_FLAG && (
+        <div id="fancy-enhanced-feature"></div>
+      )}
+
+      ...
+    </section>
+  )
+}
+```
+
+Some times you still want to read a variable that is not defined in the schema. In these cases `getSecret` is available to read runtime variables and works with any runtime (Cloudflare, Node.js, Deno).
```

Comment:
I believe here the correct would be Sometimes (one word), also, we are being a bit inconsistent with including the parenthesis for function names referenced in the text, I think we should use it for all cases.
```suggestion
Sometimes, you still want to read a variable that is not defined in the schema. In these cases, `getSecret()` is available to read runtime variables and works with any runtime (Cloudflare, Node.js, Deno).
```